crates/interpreter/src/bit_field.rs

```diff
@@ -0,0 +1,81 @@
+// Copyright 2024 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use crate::error::*;
+
+pub fn into_signed_field(mask: u32, value: i32) -> Result<u32, Error> {
+    let value = value.wrapping_add(offset(mask)) as u32;
+    into_field(mask, value)
+}
+
+pub fn from_signed_field(mask: u32, field: u32) -> i32 {
+    from_field(mask, field) as i32 - offset(mask)
+}
+
+pub fn offset(mask: u32) -> i32 {
+    1 << (mask.count_ones() - 1)
+}
+
+pub fn into_field(mask: u32, value: u32) -> Result<u32, Error> {
+    let field = (value << mask.trailing_zeros()) & mask;
+    if from_field(mask, field) != value {
+        return Err(unsupported(if_debug!(Unsupported::SideTable)));
+    }
+    Ok(field)
+}
+
+pub fn from_field(mask: u32, field: u32) -> u32 {
+    (field & mask) >> mask.trailing_zeros()
+}
+
+#[cfg(test)]
+mod tests {
+    use alloc::format;
+
+    use proptest::prelude::*;
+
+    use super::*;
+    use crate::Error;
+
+    proptest! {
+        #[test]
+        fn signed_field_round_trip(expected in -8i32..8) {
+            let mask = 0b1111000;
+            let actual = from_signed_field(mask, into_signed_field(mask, expected).unwrap());
+            prop_assert_eq!(actual, expected);
+        }
+
+        #[test]
+        fn signed_field_error_positive(expected in 8i32..100) {
```

Comment:
You may wonder why I wrote `8i32..100` instead of `8i32..` (and similarly below). I find that writing `0i32..` would actually make the test pass! The reason may be that `proptest` randomly samples values and the interval `[0, 7)` somehow was not sampled. So I use `100` to represent infinity, so that every value in `[8, 100)` are all checked.

Reply:
I see. I somehow thought proptest would be "smart" and try edge cases then random uniform. But apparently it only does random uniform. And thinking about it more, I think we should also randomize the mask if we want to do it properly.

So I see 2 options:
- We go full proptest and also randomize the mask (see below for hints)
- We go full unittest and brute-force success using the masks you chose and use edge cases for errors

To randomize a mask, you can use the `number-encoding` crate with `prop_compose!()` macro. Something along those lines:

```rust
prop_compose! {
    // We take as input all subsets of 2 distinct elements between 0 and 32 (inclusive), except for
    // the last subset, which is {31, 32}.
    fn arb_mask()(n in 0 .. combination(33, 2) - 1) -> u32 {
        let mut b = [0, 0];
        combinadics::decode_mut(n, 2, &mut b);
        // We use the smallest element of the set as the mask length minus one, since it can range
        // from 0 to 30 (we skip {31, 32}).
        let len = b[0] + 1;
        // We use the largest element of the set as one past the highest bit of the mask, since it
        // can range from 1 to 32.
        let off = b[1] - len;
        (1 << len) - 1 << off
    }
}
```

You can then add the following helper functions:

```rust
fn max_unsigned(mask: u32) -> u32 {
    (1 << mask.count_ones()) - 1
}
fn max_signed(mask: u32) -> i32 {
    (1 << mask.count_ones() - 1) - 1
}
fn min_signed(mask: u32) -> i32 {
    max_signed(mask) - max_unsigned(mask) as i32
}
```

Then proptest success with `(mask, expected) in arb_mask().prop_flat_map(|mask| (Just(mask), 0..=max_unsigned(mask)))` for unsigned and `min_signed(mask)..=max_signed(mask)` for signed.

And proptest failures with `mask in arb_mask()` and testing `max_unsigned(mask) + 1`, `min_signed(mask) - 1`, and `max_signed(mask) + 1`.

Reply:
Indeed, it is a great idea to randomize the mask in the meanwhile. The approach I take is different: making use of `prop_oneof!`. Note that the number of test runs is [256 by default](https://docs.rs/proptest/latest/proptest/test_runner/struct.Config.html). The probability that one of the three cases is chosen is `1/3`. So the probability that the edge cases are not chosen is extremely close to `0`.

Reply:
Is the branch up-to-date? I don't see using arbitrary masks, which would really be what justifies property testing, otherwise a simple unittest with a for-loop would be faster and simpler.

Using `prop_oneof` is a good idea to skew the odds, but given proptest is not smart (just random), we will waste compute testing the same thing over and over. When the search space is small, we should simply use an exhaustive approach with unittest. When it is bigger but not unreasonably big, we should use proptesting (because the search is stupid). When it is excessively big and the search needs to be smart, then fuzzing is needed.

Here, there are 527 masks (good for proptesting). And there are 12'884'901'820 pairs of mask and value (almost needs fuzzing at that point) ignoring signedness (multiply by 2 otherwise).

So I'm less and less convinced about using proptest for this usecase. If we want to test properly we would need fuzzing. Otherwise if we just want to test a single mask, then unittest is enough. Fuzzing is actually rather easy to do in Rust if you want to try it.

Reply:
PTAL at the current version. Basically I revert the tests to the earliest version in this thread with one modification: I set `PROPTEST_CASES` to be 1, which makes the tests only executed once.

The main reason for me to use `proptest` is to avoid writing for loop within a unit test, which should be handled by a testing framework ideally. The reason why I prefer the current version to the randomizing masks approach is that it is easier to read, and the latter may require additional tests for `arb_mask()` to be rigorous.

Reply:
> > PTAL at the current version.
> 
> This looks much worse. Now instead of unreliably wasting resources, it unreliably doesn't test what it should.
> 
> > I set `PROPTEST_CASES` to be 1, which makes the tests only executed once.
> 
> How is that a good thing? We want to execute each value once, not the whole proptest once. Because now there's only 1/16 chance of testing any of the values.
>

Sorry I'm confused. Just wanted to make sure we are on the same page. Regardless of `PROPTEST_CASES`, I thought `signed_field_error_positive(expected in 8..100)` means all values of `expected` from 8 to 99 are checked, right? (I did not use `prop_oneof!`.)`PROPTEST_CASES` is about the number of times `signed_field_error_positive` is executed (if it is correct). Let me know what went wrong if any. Thanks.

Reply:
And... A drawback of manually writing a for loop with `cargo test` is that it does not tell us the minimum failing value in the error message, besides we'd have to write it manually.

Reply:
> Sorry I'm confused.

That's exactly the problem of using frameworks needlessly. It adds more problems than it solves.

> Regardless of `PROPTEST_CASES`, I thought `signed_field_error_positive(expected in 8..100)` means all values of `expected` from 8 to 99 are checked, right?

No, it means it will test 256 (or whatever is `PROPTEST_CASES`) random values (maybe multiple times the same) in the `8..100` range. In particular, if you change the range to `7..100`, you will see that it won't always fail. (Once it fails it will continue failing, but it's just because proptest creates a local regression corpus. If you delete it or don't save it, you're back to square one.)

> `PROPTEST_CASES` is about the number of times `signed_field_error_positive` is executed

No, `PROPTEST_CASES` is how many random values are taken from the strategy.

Reply:
> And... A drawback of manually writing a for loop with `cargo test` is that it does not tell us the minimum failing value in the error message, besides we'd have to write it manually.

That's not really an issue. There's 2 ways to deal with that:
- Either don't care because tests are not supposed to fail. If it fails, then it's fine to lose some time debugging with a single printf.
- Or preemptively print the value of the iteration in the assert message. If the message is not too long that's fine.